bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BulkReadVeneerApi.java

```diff
@@ -0,0 +1,129 @@
+import com.google.bigtable.v2.RowFilter;
+import com.google.cloud.bigtable.data.v2.BigtableDataClient;
+import com.google.cloud.bigtable.data.v2.models.Filters;
+import com.google.cloud.bigtable.data.v2.models.Row;
+import com.google.cloud.bigtable.hbase.adapters.Adapters;
+import com.google.cloud.bigtable.hbase.wrappers.BulkReadWrapper;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.ByteString;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * Adapter for sending batch reads.
+ *
+ * <p>This class works with {@link com.google.cloud.bigtable.hbase.BatchExecutor} to enable bulk
+ * reads from the hbase api.
+ *
+ * <p>This class is not thread safe. It must be used on a single thread
+ */
+public class BulkReadVeneerApi implements BulkReadWrapper {
+
+  private final BigtableDataClient client;
+  private final String tableId;
+  private final Map<RowFilter, Batcher<ByteString, Row>> batchers;
+
+  // TODO: remove this once gax-java's Batcher supports asyncClose(). This will eliminate the need
+  private final AtomicLong cleanupBarrier;
+
+  BulkReadVeneerApi(BigtableDataClient client, String tableId) {
+    this.client = client;
+    this.tableId = tableId;
+
+    this.batchers = new HashMap<>();
+    this.cleanupBarrier = new AtomicLong();
+    this.cleanupBarrier
+        .incrementAndGet(); // wait for flush to signal before cleaning up the batcher map
+  }
+
+  @Override
+  public ApiFuture<Result> add(ByteString rowKey, @Nullable Filters.Filter filter) {
+    cleanupBarrier.incrementAndGet();
+
+    ApiFuture<Row> rowFuture = getOrCreateBatcher(filter).add(rowKey);
+    rowFuture.addListener(
+        new Runnable() {
+          @Override
+          public void run() {
+            notifyArrival();
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    return ApiFutures.transform(
+        rowFuture,
+        new ApiFunction<Row, Result>() {
+          @Override
+          public Result apply(Row row) {
+            return Adapters.ROW_ADAPTER.adaptResponse(row);
+          }
+        },
+        MoreExecutors.directExecutor());
+  }
+
+  private void notifyArrival() {
+    if (cleanupBarrier.decrementAndGet() == 0) {
+      cleanUp();
+    }
+  }
+
+  /** close all outstanding Batchers */
+  private void cleanUp() {
+    for (Batcher<ByteString, Row> batcher : batchers.values()) {
+      try {
+        batcher.close();
+      } catch (Throwable ignored) {
+        // Ignored
+      }
+    }
+    batchers.clear();
+  }
+
+  /*
+   * If autoflush is off, elementThreshold is 100, and size is 20MB, Now
+   *  -> user calls Table#batch with 10 entries
+   *     -> none of
+   */
+  @Override
+  public void flush() {
+    for (Batcher<ByteString, Row> batcher : batchers.values()) {
+      batcher.sendOutstanding();
+    }
+    notifyArrival();
+  }
```

Comment:
I will remove the code comment, The reason for this batcher.sendOutstanding is if user has switch-off autoflush then it may not resolve.

Reply:
I dont think the comment is necessary